### ISSUE TITLE
Docs: How to use avro + schema registry

### ISSUE THREAD

**Issue by powersemmi** [airtai/faststream#1774]
It seems to me that the documentation lacks an example of using avro with confluence scheme registry. I will attach a working example.

```python
import io
import json
import logging
from typing import Any, Awaitable, Callable

import fastavro
from confluent_kafka.schema_registry import SchemaRegistryClient
from faststream import FastStream
from faststream.confluent import KafkaMessage, KafkaBroker
from pydantic import BaseModel

logger = logging.getLogger(__name__)
registry = SchemaRegistryClient({'url': "http://localhost:8081"})


def avro_encoder(msg: dict, topic: str) -> bytes:
    """
    To match the kafka wire-format,
    add 1 magic byte + 4 bytes of the schema version from the registry at the beginning of the message
    https://docs.confluent.io/platform/current/schema-registry/fundamentals/serdes-develop/index.html#wire-format
    """
    subject_name = f'{topic}-value'
    value_schema = registry.get_latest_version(subject_name)
    schema_version = value_schema.version
    schema = json.loads(value_schema.schema.schema_str)
    bytes_writer = io.BytesIO()
    fastavro.schemaless_writer(fo=bytes_writer, schema=schema, record=msg)
    return b'\x00' + schema_version.to_bytes(4, byteorder='big') + bytes_writer.getvalue()


def avro_decoder(msg: bytes, topic: str) -> dict:
    """
    We pull the schema id out of the message and use it to get the actual schema
    """
    scheme_version = int.from_bytes(msg[1:5], byteorder='big', signed=False)
    value_schema = registry.get_version(f'{topic}-value', scheme_version)
    schema = json.loads(value_schema.schema.schema_str)
    return fastavro.schemaless_reader(fo=io.BytesIO(msg[5:]), writer_schema=schema)


def decoder_factory(topics: tuple[str]) -> Callable[[KafkaMessage], dict]:
    def decode_message(msg: KafkaMessage) -> dict:
        for topic in topics:
            try:
                return avro_decoder(msg=msg.body, topic=topic)
            except Exception as e:
                logger.info('failed parce message with err=%s', e)
        raise ValueError(f'Failed parce message {msg}')

    return decode_message


def publisher_encoder_factory(topic: str) -> Callable[[Any], Awaitable[Any]]:
    async def encode_message(call_next: Callable[..., Awaitable[Any]], msg: BaseModel | dict, **options: Any) -> Any:
        match msg:
            case BaseModel():
                msg_to_process = msg.model_dump(mode='json')
            case dict():
                msg_to_process = json.dumps(msg)
            case _:
                raise ValueError(f'{type(msg)=} is type not in (BaseModel, dict)')
        return await call_next(avro_encoder(msg=msg_to_process, topic=topic), **options)

    return encode_message


broker = KafkaBroker()
app = FastStream(broker)

publisher = broker.publisher(
    topic='test',
    middlewares=[publisher_encoder_factory(topic='test')],
)


class Consumer(BaseModel):
    name: str
    age: int


@broker.subscriber("test", decoder=decoder_factory(topics=("test",)))
async def consume(body: Consumer):
    logger.info("name=%s, age=%s", body.name, body.age)


@app.after_startup
async def publish():
    msg = {"name": "John", "age": 25}

    await publisher.publish(msg)

```
Avro scheme sample
```json
{
  "fields": [
    {
      "name": "payload",
      "type": {
        "fields": [
          {
            "name": "name",
            "type": "string"
          },
          {
            "name": "age",
            "type": "int"
          }
        ],
        "name": "payload_record",
        "type": "record"
      }
    }
  ],
  "name": "test",
  "namespace": "test",
  "type": "record"
}
```

**Comment by Lancetnik** [comment]
I think, we should please such reciepe in [Kafka How-To](https://faststream.airt.ai/latest/howto/kafka/) section
I schedule this task, but we are pretty busy with a code, so any contribution are welcome
Thank you for the investigation and code example a lot! You made a great work!

**Comment by Lancetnik** [comment]
> @Lancetnik And it seems to me that an example can be an addition to [serialization/examples](https://faststream.airt.ai/latest/getting-started/serialization/examples/#avro)

We can just add link to this section due your example is Confluent-specific

**Comment by powersemmi** [comment]
> I found an interesting project: https://github.com/mlovretovich/faststream-schema-registry Looks like the easiest way to support the feature

"I think this is not a very popular solution, and using it on a production server would be risky," the team leader is quoted as saying.

**Comment by ArtemIlinArammeem** [comment]
@powersemmi thank you for your example. I think the if/else condition in `confluent_encoder` should be inverted though. If the schema already exists in cache, no point in calling the registry.

**Comment by powersemmi** [comment]
TL;DR. Here is the updated version of the encoder/decoder

```python
import io
import json
import logging
from collections.abc import Iterable
from dataclasses import dataclass
from decimal import Decimal
from enum import Enum
from typing import Any, Literal, TypeAlias

import fastavro
from cachetools import TTLCache
from confluent_kafka import cimpl
from confluent_kafka.schema_registry import SchemaRegistryClient, SchemaRegistryError
from fastavro._validation import ValidationError  # type: ignore[attr-defined]
from faststream import BaseMiddleware
from faststream.confluent import KafkaMessage
from pydantic import BaseModel

from app.config import config

logger = logging.getLogger(__name__)

registry = SchemaRegistryClient({'url': config.KAFKA_SCHEMA_REGISTRY_ADDRESS})

schema_cache = TTLCache(maxsize=128, ttl=60 * 10)  # type: ignore[var-annotated]

DecodedMessage: TypeAlias = str | float | int | Decimal | bytes | list[Any] | dict[Any, Any] | None


# Supported schema types
class SchemaTypeEnum(Enum):
    AVRO = 'AVRO'
    JSON = 'JSON'


@dataclass
class Schema:
    schema_id: int
    schema_obj: dict[str, Any]
    schema_type: SchemaTypeEnum


class DecoderExceptionError(Exception):
    pass


class EncoderExceptionError(Exception):
    pass


def _get_schema_from_registry(subject_name: str) -> Schema:
    """
    Gets a schema from Schema Registry and updates the cache.
    """
    try:
        registry_schema = registry.get_latest_version(subject_name)
        schema_id = registry_schema.schema_id
        schema_type = SchemaTypeEnum(registry_schema.schema.schema_type)
        schema_obj = json.loads(registry_schema.schema.schema_str)
        schema = Schema(schema_id=schema_id, schema_obj=schema_obj, schema_type=schema_type)
        schema_cache[subject_name] = schema
        logger.info('Schema for subject %s loaded with ID %s', subject_name, schema_id)
        return schema
    except SchemaRegistryError as e:
        logger.error('Failed to fetch schema for subject %s: %s', subject_name, e)
        raise


def get_schema(subject_name: str) -> Schema:
    """
    Returns schema from cache or registry.
    """
    schema_name = schema_cache.get(subject_name)
    if schema_name is None:
        return _get_schema_from_registry(subject_name)
    return schema_cache[subject_name]


def avro_encoder(msg: DecodedMessage, schema_obj: dict[str, DecodedMessage]) -> bytes:
    """
    Encodes a message in Avro format.
    """
    bytes_writer = io.BytesIO()
    fastavro.schemaless_writer(bytes_writer, schema_obj, msg)
    return bytes_writer.getvalue()


ENCODERS = {
    SchemaTypeEnum.AVRO: avro_encoder,
    SchemaTypeEnum.JSON: lambda msg, _: json.dumps(msg).encode(),
}


def confluent_encoder(msg: DecodedMessage, topic: str, subject_type: Literal['value', 'key'] = 'value') -> bytes:
    """
    Encodes a message in wire-format for Kafka.
    """
    subject_name = f'{topic}-{subject_type}'
    schema = get_schema(subject_name)
    encoder = ENCODERS.get(schema.schema_type)
    if not encoder:
        raise EncoderExceptionError(f'Unsupported schema type: {schema.schema_type}')
    try:
        fastavro.validation.validate(msg, schema.schema_obj)
    except ValidationError as e:
        raise EncoderExceptionError(f'Message does not conform to schema: {e}') from e
    encoded_data = encoder(msg, schema.schema_obj)
    return b'\x00' + schema.schema_id.to_bytes(4, byteorder='big') + encoded_data


def _confluent_decoder(body: cimpl.Message) -> DecodedMessage:
    """
    Decodes a message from Kafka wire-format.
    """
    value = body.value()
    if len(value) < 5 or value[0] != 0x00:
        raise DecoderExceptionError('Invalid wire-format message')

    schema_id = int.from_bytes(value[1:5], byteorder='big')
    msg_data = value[5:]

    try:
        schema = registry.get_schema(schema_id)
    except SchemaRegistryError as e:
        logger.error('Failed to fetch schema with ID %s: %s', schema_id, e)
        raise

    schema_type = SchemaTypeEnum(schema.schema_type)
    schema_obj = json.loads(schema.schema_str)

    if schema_type == SchemaTypeEnum.AVRO:
        return fastavro.schemaless_reader(io.BytesIO(msg_data), schema_obj)  # type: ignore[call-arg]
    else:
        raise DecoderExceptionError(f'Unsupported schema type: {schema.schema_type}')


def confluent_decoder(msg: KafkaMessage) -> DecodedMessage:
    if isinstance(msg.raw_message, Iterable):
        return [_confluent_decoder(m) for m in msg.raw_message]
    return _confluent_decoder(msg.raw_message)


class AvroEncoderMiddleware(BaseMiddleware):
    async def on_publish(self, msg: Any, **kwargs: Any) -> bytes:
        topic = kwargs['topic']
        match msg:
            case BaseModel():
                msg_to_process = msg.model_dump(mode='python')
            case str():
                msg_to_process = json.loads(msg)
            case dict():
                msg_to_process = msg
            case _:
                raise ValueError(f'Unsupported message type: {type(msg)}')

        encoded_message = confluent_encoder(msg_to_process, topic)
        return encoded_message

```


Using:

```python
from faststream.confluent import KafkaBroker

broker = KafkaBroker(
    bootstrap_servers=config.KAFKA_SERVERS,
    middlewares=[AvroEncoderMiddleware],
    decoder=confluent_decoder,
    graceful_timeout=60.0 * 10,
)

...
```

**Comment by Lancetnik** [comment]
#1297 (docs for different schema registries) is closed as a duplicate of this one. When this gets written up, the page should keep the registry behind a plain decoder/encoder pair, as in the example above, so that the other registries fit the same pattern.
